Improve bench-tps keypair generation

jstarry opened the pull request:
#### Problem
Account generation at the beginning of bench-tps can sometimes be really slow. This could be due to a number of reasons:
1. If a large portion of a batch of transfers fails, generation has no way to bail out of attempting to verify each transfer.
2. RPC requests are slow due to geographic latency

#### Summary of Changes
* Parallelize fund verification so that RPC requests are less of a bottleneck
* Short circuit fund verification if more than half of the tx's failed
* Short circuit verification of a tx if any of the "to" keypairs is funded or not funded
* Transfer direction  changes more frequently so that source/dest keypair funds stay balanced
* Made the initial funds check much more lenient so that we can avoid re-running fund generation
* General code cleanup

Fixes: https://github.com/solana-labs/solana/issues/7597

jstarry commented:
@sakridge @mvines this is ready for review now

mvines commented:
@sakridge has a better eye than I do in for this part of the code, I defer review to him

sakridge commented on bench-tps/src/bench.rs:
```diff
@@ -176,7 +168,7 @@ where
     let keypair_chunks = source_keypair_chunks.len() as u64;
     let mut reclaim_lamports_back_to_source_account = false;
-    let mut i = keypair0_balance;
+    let mut i = 0;
```
why change this?

jstarry replied:
So that mechanism was in place to quickly restart `bench-tps` with the same parameters as a previous run. We would detect the balance of the last keypair to check how many lamports had been transferred from one half of they keypairs to the other half. But I found that the assumptions were not valid for running with `ramp-tps`.

In this PR, I removed the quick-start mechanism but I think that was a mistake. I'll think of another approach that works well for repeated `bench-tps` runs as well as incremental `ramp-tps` runs

sakridge commented on bench-tps/src/bench.rs:
```diff
@@ -625,42 +586,73 @@ pub fn fund_keys<T: Client>(
-                        if timer.elapsed() >= Duration::from_secs(5) {
-                            if failed_verify > 0 {
-                                debug!("total txs failed verify: {}", failed_verify);
+                    let failed_verify = Arc::new(AtomicUsize::new(0));
```
what do you think about breaking this up into another function. The complexity was bad before, now it's even worse.

jstarry replied:
Totally in favour, I'll clean this up

sakridge commented:
Overall question, why are the funding transactions failing in the original?

jstarry commented:
@sakridge I updated the PR and it's ready for another pass. I updated the logic around transfer direction to be a lot simpler. I tested the changes briefly on a cpu testnet and behaviour looks good. I'll run with gpus later today to make sure tps isn't affected.

sakridge commented:
> @sakridge I updated the PR and it's ready for another pass. I updated the logic around transfer direction to be a lot simpler. I tested the changes briefly on a cpu testnet and behaviour looks good. I'll run with gpus later today to make sure tps isn't affected.

ok cool, i'll take a look

sakridge commented on bench-tps/src/bench.rs:
```diff
@@ -1004,63 +1031,70 @@ fn fund_move_keys<T: Client>(
+    // 100 lamports should give enough wiggle room to handle source / dest keypair sets getting unbalanced
+    let minimum_balance = 100;
+
+    if first_keypair_balance < minimum_balance || last_keypair_balance < minimum_balance {
```
I'm not sure I get this minimum balance check fixed at 100. The user could have specified 1000 or 10,000 lamports per account, then if they have only 100 they would not be funded, is that correct? The cluster fees can be in the 1000 lamport range also, so I'm not sure a fixed low value here will work in all cases.

jstarry replied:
@sakridge you're right, this was oversimplified and didn't take tx fees into account. I updated the PR with a new approach: https://github.com/solana-labs/solana/pull/7723/commits/78516ca48db91cdbcfe79b3b314c11e8ecd96178

jstarry commented:
I wasn't able to get a gpu testnet working but ran a gce cpu testnet with 25k avg tps and tx errors were pretty quiet so I feel pretty confident in this new approach of switching transfer directions more frequently.
